rust/agama-dbus-server/src/network/model.rs

```diff
@@ -306,24 +306,96 @@ impl Connection {
     pub fn is_loopback(&self) -> bool {
         matches!(self, Connection::Loopback(_))
     }
+
+    pub fn is_ethernet(&self) -> bool {
+        matches!(self, Connection::Loopback(_)) || matches!(self, Connection::Ethernet(_))
+    }
 }
 
 #[derive(Debug, Default, Clone)]
 pub struct BaseConnection {
     pub id: String,
     pub uuid: Uuid,
+    pub mac_address: Option<MacAddress>,
     pub ip_config: IpConfig,
     pub status: Status,
     pub interface: String,
     pub match_config: MatchConfig,
 }
 
+impl BaseConnection {
+    pub fn mac_address_string(&self) -> String {
+        match &self.mac_address {
+            Some(mac) => mac.to_string(),
+            None => "".to_string(),
+        }
+    }
+
+    pub fn set_mac_address(&mut self, mac_address: &str) -> Result<(), InvalidMacAddress> {
+        self.mac_address = if mac_address.is_empty() {
+            None
+        } else {
+            Some(MacAddress::from_str(mac_address)?)
+        };
+        Ok(())
+    }
+}
+
 impl PartialEq for BaseConnection {
     fn eq(&self, other: &Self) -> bool {
         self.id == other.id && self.uuid == other.uuid && self.ip_config == other.ip_config
     }
 }
 
+#[derive(Debug, Error)]
+#[error("Invalid MAC address: {0}")]
+pub struct InvalidMacAddress(String);
+
+#[derive(Debug, Clone)]
+pub enum MacAddress {
+    MacAddress(macaddr::MacAddr6),
+    Preserve,
+    Permanent,
+    Random,
+    Stable,
+}
+
+impl FromStr for MacAddress {
+    type Err = InvalidMacAddress;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "preserve" => Ok(Self::Preserve),
+            "permanent" => Ok(Self::Preserve),
+            "random" => Ok(Self::Permanent),
```

Comment:
Why rare `"random"` and `"stable"` converted to `Permanent`?

Reply:
Oops must have been some lazy copy pasting :smile: